core/genesis/files_test.go

```diff
@@ -0,0 +1,408 @@
+	m.Value = "value"
+
+	bz, err := proto.Marshal(m)
+	require.NoError(t, err)
+
+	_, err = f.Write(bz)
+	require.NoError(t, err)
+
+	gs := NewFileGenesisSource(tmpdir, moduleName)
+
+	mr := &testpb.TestGenesisFile{}
+
+	err = gs.ReadMessage(mr)
+	require.NoError(t, err)
+
+	require.Equal(t, m.String(), mr.String())
+}
+
+func TestFileGenesisTargetWithField(t *testing.T) {
+	tmp := t.TempDir()
+
+	moduleName := ""
+	field := "field"
+
+	// empty module name, should return error
+	gs := NewFileGenesisTarget(tmp, moduleName)
+	_, err := gs.OpenWriter(field)
+	require.Error(t, err)
+
+	// open file with module field
+	moduleName = "module"
+	gs = NewFileGenesisTarget(tmp, moduleName)
+	w, err := gs.OpenWriter(field)
+	require.NoError(t, err)
+	defer w.Close()
+
+	msg := "test message"
+	_, err = w.Write([]byte(msg))
+	require.NoError(t, err)
+
+	fp, err := os.Open(filepath.Join(tmp, moduleName, fmt.Sprintf("%s.json", field)))
+	require.NoError(t, err)
+	defer fp.Close()
+
+	buf := &bytes.Buffer{}
+	_, err = buf.ReadFrom(fp)
+	require.NoError(t, err)
+
+	require.Equal(t, msg, buf.String())
+}
+
+func TestFileGenesisTargetWithModule(t *testing.T) {
+	tmp := t.TempDir()
+
+	moduleName := "test"
+	field := ""
+
+	gs := NewFileGenesisTarget(tmp, moduleName)
+	w, err := gs.OpenWriter(field)
+	require.NoError(t, err)
+	defer w.Close()
+
+	msg := "test message"
+	_, err = w.Write([]byte(msg))
+	require.NoError(t, err)
+
+	fp, err := os.Open(filepath.Join(tmp, fmt.Sprintf("%s.json", moduleName)))
+	require.NoError(t, err)
+	defer fp.Close()
+
+	buf := &bytes.Buffer{}
+	_, err = buf.ReadFrom(fp)
+	require.NoError(t, err)
+
+	require.Equal(t, msg, buf.String())
+}
+
+func TestFileGenesisTargetWithoutModuleAndField(t *testing.T) {
+	tmp := t.TempDir()
+
+	moduleName := ""
+	field := ""
+
+	gs := NewFileGenesisTarget(tmp, moduleName)
+	w, err := gs.OpenWriter(field)
+	require.NoError(t, err)
+	defer w.Close()
+
+	msg := "test message"
+	_, err = w.Write([]byte(msg))
+	require.NoError(t, err)
+
+	fp, err := os.Open(filepath.Join(tmp, "genesis.json"))
+	require.NoError(t, err)
+	defer fp.Close()
+
+	buf := &bytes.Buffer{}
+	_, err = buf.ReadFrom(fp)
+	require.NoError(t, err)
+
+	require.Equal(t, msg, buf.String())
+}
+
+func TestFileGenesisTargetWriteMessageNoOp(t *testing.T) {
+	var pm proto.Message
+	gs := NewFileGenesisTarget("", "")
+	err := gs.WriteMessage(pm)
+	require.Error(t, err, "unsupported op")
+}
+
+func TestFileGenesisTargetWriteRawJSONEmptyModuleName(t *testing.T) {
+	tmpdir := t.TempDir()
+	moduleName := ""
+
+	gt := NewFileGenesisTarget(tmpdir, moduleName)
+
+	str := `{"genesis": "write test"}`
+	err := gt.WriteRawJSON([]byte(str))
+	require.ErrorContains(t, err, "failed to write RawJSON: empty module name")
+}
+
+func TestFileGenesisTargetWriteRawJSON(t *testing.T) {
+	tmpdir := t.TempDir()
+	moduleName := "module"
+
+	gt := NewFileGenesisTarget(tmpdir, moduleName)
+
+	str := `{"genesis": "write test"}`
+	err := gt.WriteRawJSON([]byte(str))
```

Comment:
A sign that you should be using []byte as is and even down below with the simplified changes.

```go
in := []byte(`{"genesis": "write test"}`)
err := gt.WriteRawJSON(in)
```